### PR TITLE
fabric-mining-levels-v0 -> fabric-tool-attribute-api-v1 (1.15 edition)

### PR DESCRIPTION
**Pull request by LemmaEOF**
Reimplementation of #352 on the 1.15 branch, now without merge conflicts!

Moves the ToolManager and tool tags declared in fabric-mining-levels-v0 to a new fabric-tool-attributes-v1, along with adding new interfaces to be implemented on tools so that mining speed, mining level, and entity attribute modifiers can be decided by the itemstack.

### CODE REVIEW

**Comment by LemmaEOF** [comment]
I have confirmed this builds successfully and has no checkstyle problems.

File: fabric-tool-attributes-v1/build.gradle
Comment (liach) [inline fabric-tool-attributes-v1/build.gradle]: Afaik this breaks java 9 and above. Consider using something else like checker qual or jetbrain annotations.

Comment (LemmaEOF) [reply fabric-tool-attributes-v1/build.gradle]: Oh boy, it does seem like JSR305 got busted in JDK9. I could maybe use Spotbugs, which is a continuation of Findbugs. I'm not sure how much I like the concept of using IDE-specific annotations.

Comment (LemmaEOF) [reply fabric-tool-attributes-v1/build.gradle]: Moved to Spotbugs.

File: fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/ToolActor.java
Comment (i509VCB) [inline fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/ToolActor.java]: Do we have a consensus anywhere on which Nullable we are even going to use? I remember one discussion where people seemed to agree something like jetbrains was the preferred option but no firm decision from what I see anywhere else in the codebase.

File: fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/ToolAttributeHolder.java
Comment (i509VCB) [inline fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/ToolAttributeHolder.java]: Maybe a MiningLevelConstants/MiningLevels class which contains the default vanilla mining levels all neatly laid out?

File: fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/ToolAttributeHolder.java
Comment (i509VCB) [inline fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/ToolAttributeHolder.java]: Similar to the comment above, maybe a constants class for the default vanilla values.

File: fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/AttributeManager.java
Comment (i509VCB) [inline fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/AttributeManager.java]: 🍫

**Review by i509VCB — COMMENTED** [review]
Really just two recommendations for convince, I don't see anything wrong from my glance.

**Comment by i509VCB** [comment]
Whoever runs this place, please tag this PR.

File: fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/ToolActor.java
Comment (LemmaEOF) [reply fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/ToolActor.java]: I think this might be the first Nullable in Fabric API, actually.

File: fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/ToolAttributeHolder.java
Comment (LemmaEOF) [reply fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/ToolAttributeHolder.java]: I think that might be nice. I'll think about it when I have time.

File: fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/AttributeManager.java
Comment (LemmaEOF) [reply fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/AttributeManager.java]: ?

Comment (Pyrofab) [reply fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/AttributeManager.java]: Manningham Mills can add values to enum

Comment (LemmaEOF) [reply fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/AttributeManager.java]: oh god right, Manningham Mills can add values to enums. I'll remove this throw and just have it no-op.

File: fabric-mining-levels-v0/src/main/resources/fabric.mod.json
Comment (Player3324) [inline fabric-mining-levels-v0/src/main/resources/fabric.mod.json]: "..Deprecated and replaced by .." please

File: fabric-tool-attributes-v1/build.gradle
Comment (Player3324) [inline fabric-tool-attributes-v1/build.gradle]: After some further consideration I am convinced we should use singular for module names to have a perfect match with the package. So fabric-tool-attribute-v1.

File: fabric-tool-attributes-v1/build.gradle
Comment (Player3324) [inline fabric-tool-attributes-v1/build.gradle]: I'd postpone the addition of static analysis tooling to a separate discussion/changeset.

File: fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/ActableAttributeHolder.java
Comment (Player3324) [inline fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/ActableAttributeHolder.java]: The name is not very descriptive, technically any tool is "actable".

Overloads in a context where users both call and implement are asking for bugs around missing forwards or needlessly calling the generic variant. If feasible I'd like to see only the extensive version, it also simplifies the API.

ToolActor is a rather weird box, the generic parameter isn't constrained anywhere I can tell and the API is merely `box()`+`unbox()`. This makes it almost as universal as using Object directly while still having to deal with wrapping. There are a few alternatives:
- Use LivingEntity instead (probably the best with fake players living in this hierarchy)
- Use Object instead (as useless to implementers as ToolActor)
- Use a "property accessor" exposing all actor properties of interest (probably not practical)
- Use a container object carrying a fixed variety of known object types (e.g. block entity + tile entity - probably not better than LivingEntity with the fake player exposing the be somehow)

File: fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/FabricToolTags.java
Comment (Player3324) [inline fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/FabricToolTags.java]: Not sure these should be plural, what is the convention elsewhere?

File: fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/ActableAttributeHolder.java
Comment (Player3324) [inline fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/ActableAttributeHolder.java]: Mining level is rather weakly specified and IIRC has been too coarse for other similar APIs elsewhere. Maybe we can remap the range and provide some guidelines, especially for harder materials? Something like `fabric level = 8 * vanilla level` and a suggestion to continue tiers in equal steps, making the tier above diamond 8 higher?

File: fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/ToolActor.java
Comment (LemmaEOF) [reply fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/ToolActor.java]: This will be changed when #461 is merged

File: fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/ActableAttributeHolder.java
Comment (Player3324) [inline fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/ActableAttributeHolder.java]: "unless you want doubling" is not very understandable.

**Comment by LemmaEOF** [comment]
All current feedback has been implemented

File: fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/ToolAttributeHolder.java
Comment (Player3324) [inline fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/ToolAttributeHolder.java]: There should be a blank line after the initial short description in Javadoc, which should be a single sentence. Further paragraphs also need to use `<p>`.

File: fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/AttributeManager.java
Comment (Player3324) [inline fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/AttributeManager.java]: D suffix is superfluous

File: fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/ToolManager.java
Comment (Player3324) [inline fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/ToolManager.java]: The previous formatting here was fine.

File: fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/MiningConstants.java
Comment (Player3324) [inline fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/MiningConstants.java]: .0F suffix is superfluous

File: fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/ActableAttributeHolder.java
Comment (Player3324) [inline fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/ActableAttributeHolder.java]: Each API class or method should at least have the one-sentence description unless it is completely obvious. In this particular case not everyone may know what mining levels are or how the speed works.

**Review by Player3324 — CHANGES_REQUESTED** [review]
The approach looks reasonable, I've noticed some areas that may be suitable for simplification.

Please add some brief example "how to use this" code in class Javadoc and reference it from other API class' Javadoc where appropriate. As usual this is not intended to replace the wiki, but it shouldn't be necessary to study the code extensively without it either.

File: fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/FabricToolTags.java
Comment (LemmaEOF) [reply fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/FabricToolTags.java]: These are the same as the current fabric-mining-levels-v0 tags, just with shears added. Minecraft typically uses plural, ie `minecraft:planks:`.

File: fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/ActableAttributeHolder.java
Comment (LemmaEOF) [reply fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/ActableAttributeHolder.java]: I'm not sure if that'd be too invasive for Fabric's taste. It should be talked over.

Comment (Player3324) [reply fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/ActableAttributeHolder.java]: I don't know how invasive that'd be. If feasible, my preferred approach here would `round(fabricLevel / 8f)` when crossing into a vanilla context or the full level when both block and tool are using fabric levels.

File: fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/ActableAttributeHolder.java
Comment (LemmaEOF) [reply fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/ActableAttributeHolder.java]: I'll try to think of a better name.

File: fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/ActableAttributeHolder.java
Comment (bluebear94) [reply fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/ActableAttributeHolder.java]: That could introduce inconsistent behaviour between vanilla and Fabric contexts (consider a FL 10 tool trying to mine a FL 11 block: this will fail with Fabric levels but pass when both are converted to vanilla levels).

**Comment by LemmaEOF** [comment]
Most feedback has been implemented. A couple more things talked over in discord:
- `ActableAttributeHolder` has been removed, and `ToolAttributeHolder` now takes a nullable `LivingEntity` arg
- `ToolAttributeHolder` should probably be renamed, it's not really holding stuff
- javadocs need to be updated for noting nullability
- we might want to do some fancy stuff for tooltip support

**Comment by bluebear94** [comment]
> * `ToolAttributeHolder` should probably be renamed, it's not really holding stuff

What about `DynamicAttributeTool`?

**Comment by LemmaEOF** [comment]
`DynamicAttributeTool` might be good.

File: fabric-tool-attribute-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/ToolAttributeHolder.java
Comment (dexman545) [inline fabric-tool-attribute-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/ToolAttributeHolder.java]: Similarly to [Yarn#1069](https://github.com/FabricMC/yarn/issues/1069), `getMiningSpeed` does not return the speed of mining/breaking the block - it returns one factor of many that determine the mining speed.

Comment (LemmaEOF) [reply fabric-tool-attribute-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/ToolAttributeHolder.java]: Until Yarn#1069 is resolved, this will stay `getMiningSpeed`.

Comment (Prospector) [reply fabric-tool-attribute-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/ToolAttributeHolder.java]: IMO Fabric API should try to use the best names it can and yarn can be updated to match.

Comment (LemmaEOF) [reply fabric-tool-attribute-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/ToolAttributeHolder.java]: That's not been the standard up to this point, so that should probably be discussed.

Comment (Prospector) [reply fabric-tool-attribute-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/ToolAttributeHolder.java]: Reason being API can't change easily at all while Yarn can

**Comment by LemmaEOF** [comment]
More changes made:
- for now, `getMiningSpeed` has been changed to `getMiningSpeedMultiplier`. I still maintain we resolve #1069 before this is merged so that we can come to a proper decision on what to call it.
- `ToolAttributeHolder` has been renamed to `DynamicAttributeTool`.
- Additional javadocs have been added.
- I've fixed some code in `MixinLivingEntity` that still queried `ActableAttributeHolder`, which no longer exists.

This should fix all the feedback I remember on this, other than Player's suggestion of increasing the distance between each vanilla mining level, since I'm not 100% sure how comfortable I am with that.

File: fabric-tool-attribute-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/MiningConstants.java
Comment (shedaniel) [inline fabric-tool-attribute-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/MiningConstants.java]: Isn't this for 1.15, are we sure that we want netherite here

Comment (LemmaEOF) [reply fabric-tool-attribute-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/MiningConstants.java]: I mean it's just a named constant, so I don't think there's any harm in having it there as reference material.

File: fabric-tool-attribute-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/MiningConstants.java
Comment (i509VCB) [inline fabric-tool-attribute-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/MiningConstants.java]: Is there a way to refer to a concrete value in the game rather than constants mojang could change at any time during snapshots?

**Review by i509VCB — APPROVED** [review]
Everything looks fine to me, I'll see if I can get some more movement on an official nullable annotation.

File: fabric-tool-attribute-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/MiningConstants.java
Comment (liach) [reply fabric-tool-attribute-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/MiningConstants.java]: This would rather be impl specific. I think currently holding these in constants shouldn't be too bad, but I don't know if the javac will optimize these field calls in mods so that they get replaced by senseless literal numbers.

Comment (shedaniel) [reply fabric-tool-attribute-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/MiningConstants.java]: I wonder if we want these fields anyways, as this is just for reference, I feel like mod developers might want to just do `ToolMaterials.GOLD.<insert method to get mining speed>`

Comment (LemmaEOF) [reply fabric-tool-attribute-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/MiningConstants.java]: I put em in because Player recommended em.

Comment (shedaniel) [reply fabric-tool-attribute-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/MiningConstants.java]: If a mod changes the vanilla values with a mixin and something (maybe because of reasons), these will not be correct, I feel like we shouldn't add this just because this is a handy reference.

**Review by falseresync — APPROVED** [review]

**Comment by LemmaEOF** [comment]
MiningConstants has been removed as per feedback. @modmuss50 @sfPlayer1 this is triage-ready

**Review by Player3324 — APPROVED** [review]

**Review by liach — APPROVED** [review]
I see no problem :heavy_check_mark:

**Review by modmuss50 — APPROVED** [review]
Looks good, should be moved to last call.

**Comment by Player3324** [comment]
Please consider adding some example use code to the Javadoc in relevant API classes and reference it from the remaining API classes before last call is over or in a separate PR.

File: fabric-tool-attribute-v1/build.gradle
Comment (Player3324) [inline fabric-tool-attribute-v1/build.gradle]: Please update to 1.0.0, we don't use 0.x anymore

File: fabric-tool-attribute-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/AttributeManager.java
Comment (Earthcomputer) [inline fabric-tool-attribute-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/AttributeManager.java]: Are you sure we want to combine `MULTIPLY_BASE` and `MULTIPLY_TOTAL` like this? It seems rather odd, especially since vanilla does it differently iirc.

Comment (LemmaEOF) [reply fabric-tool-attribute-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/AttributeManager.java]: I'll double-check, but I'm pretty sure that this is how vanilla does it.

Comment (LemmaEOF) [reply fabric-tool-attribute-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/AttributeManager.java]: I'm gonna need to get a pen and paper to figure this out. It looks *pretty* similar to what vanilla does, but I wanna verify.

Comment (LemmaEOF) [reply fabric-tool-attribute-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/AttributeManager.java]: god, EAMs are so weird and hacky. There's two local vars in totaling, `d` and `e`.  It goes through these steps:
1. `d` starts at the base value of the attribute, then iterates over all the `ADDITION` modifiers, adding them each to `d`.
2. `e` is set to the value of `d`, then iterates over all the `MULTIPLY_BASE` modifiers, with the operation `e += d * modifier.getAmount()` .
3. it then iterates over all the `MUTLIPLY_TOTAL` modifiers, with the operation `e *= 1.0D + entityAttributeModifier3.getAmount()`.
4. lastly, it clamps the value based on the attribute's min and max.

I'm thinking we *might* be able to use the associative property on `MULTIPLY_BASE` but I just really can't tell for sure unless I do some pretty rigorous testing.

Comment (LemmaEOF) [reply fabric-tool-attribute-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/AttributeManager.java]: alright, testing proves this doesn't work. I'm not sure how to actually *make* it work, is the fundamental issue.

Comment (LemmaEOF) [reply fabric-tool-attribute-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/AttributeManager.java]: I've gotten `MULTIPLY_TOTAL` to work properly, but I'm not sure it's even *possible* to get `MULTIPLY_BASE` to work, due to the operations it performs. My two options are to run low or run high. Which do you think is better?

Comment (LemmaEOF) [reply fabric-tool-attribute-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/AttributeManager.java]: holy shit, I found a way to make the math work perfectly!

**Comment by LemmaEOF** [comment]
It's been a week since last call started, and I'm gonna do a separate PR for javadocs since I haven't had time for them this week.

**Comment by liach** [comment]
player said as javadocs doesn't affect compatibility, it is safe to change without last call checks

File: fabric-tool-attributes-v1/build.gradle
Comment (liach) [reply fabric-tool-attributes-v1/build.gradle]: Imo we should stick with singular names.

Comment (LemmaEOF) [reply fabric-tool-attributes-v1/build.gradle]: `fabric-tool-attribute-v1` makes no sense, and it's Java code style to use plurals where necessary. It's necessary here, because each tool has multiple attributes.

**Review by Ueaj-Kerman — APPROVED** [review]

**Comment by LemmaEOF** [comment]
Module name changed again for the new scheme.

File: fabric-mining-levels-v0/src/main/resources/fabric.mod.json
Comment (i509VCB) [inline fabric-mining-levels-v0/src/main/resources/fabric.mod.json]:
```suggestion
    "fabric-tool-attribute-api-v1": "*"
```
Since the `fabric.mod.json` in v1 resources uses the new name.

**Review by i509VCB — APPROVED** [review]
Also @Boundarybreaker please give the client and server a test run in dev to make sure the mod loads on both before merging

File: fabric-mining-levels-v0/src/main/resources/fabric.mod.json
Comment (LemmaEOF) [reply fabric-mining-levels-v0/src/main/resources/fabric.mod.json]: oh fuck right, I've gotta do this on object-builders too

**Comment by LemmaEOF** [comment]
Caught an issue and fixed it

**Comment by shedaniel** [comment]
hello yes people this has been in final call for 26 days and as the new naming scheme is considered we shall merge this

**Pull request merged**
